### PR TITLE
feat(metadata-calculator): option to use VM runner for protective reads

### PR DESCRIPTION
**Pull request by itegulov**
## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/bin/external_node/src/main.rs [hunk 1]
```diff
@@ -139,6 +139,7 @@ async fn run_tree(
             .merkle_tree_include_indices_and_filters_in_block_cache,
         memtable_capacity: config.optional.merkle_tree_memtable_capacity(),
         stalled_writes_timeout: config.optional.merkle_tree_stalled_writes_timeout(),
+        sealed_batches_have_protective_reads: config.optional.protective_reads_persistence_enabled,
```
Comment (itegulov) [inline core/bin/external_node/src/main.rs:142]: I am not 100% sure about this. My thinking was:
case 1) if EN is not persisting protective reads then it is also not running tree
case 2) if EN is persisting protective reads then we can run metadata calculator with `sealed_batches_have_protective_reads=true`

@perekopskiy could you double check this please?

File: core/node/metadata_calculator/src/updater.rs [hunk 2]
```diff
@@ -184,28 +187,38 @@ impl TreeUpdater {
...
     ) -> anyhow::Result<()> {
-        let Some(last_sealed_l1_batch) = storage
-            .blocks_dal()
-            .get_sealed_l1_batch_number()
-            .await
-            .context("failed loading sealed L1 batch number")?
-        else {
-            tracing::trace!("No L1 batches to seal: Postgres storage is empty");
-            return Ok(());
+        let last_l1_batch_with_protective_reads = if self.sealed_batches_have_protective_reads {
```
Comment (perekopskiy) [inline core/node/metadata_calculator/src/updater.rs:192]: if `self.tree.mode()` is "light" then it should use `get_sealed_l1_batch_number` regardless config var

File: core/node/node_framework/examples/main_node.rs [hunk 3]
```diff
@@ -135,9 +135,11 @@ impl MainNodeBuilder {
     fn add_metadata_calculator_layer(mut self) -> anyhow::Result<Self> {
         let merkle_tree_env_config = DBConfig::from_env()?.merkle_tree;
         let operations_manager_env_config = OperationsManagerConfig::from_env()?;
+        let state_keeper_config = StateKeeperConfig::from_env()?;
```
Comment (perekopskiy) [inline core/node/node_framework/examples/main_node.rs:138]: nit: state_keeper_**env**_config for consistency

File: core/bin/zksync_server/src/node_builder.rs [hunk 4]
```diff
@@ -175,9 +175,11 @@ impl MainNodeBuilder {
         let merkle_tree_env_config = try_load_config!(self.configs.db_config).merkle_tree;
         let operations_manager_env_config =
             try_load_config!(self.configs.operations_manager_config);
+        let state_keeper_config = try_load_config!(self.configs.state_keeper_config);
```
Comment (perekopskiy) [inline core/bin/zksync_server/src/node_builder.rs:178]: nit: state_keeper_**env**_config for consistency

File: core/bin/external_node/src/main.rs (hunk 1 above)
Comment (perekopskiy) [reply core/bin/external_node/src/main.rs:142]: Yes, correct. More precisely -- EN never runs full tree, so it doesn't matter what we set in `sealed_batches_have_protective_reads`

**Review by perekopskiy — APPROVED** [review]

**Pull request merged**